### PR TITLE
feature added to avoid joint limit with ComputeJointVelocityFromTwist

### PR DESCRIPTION
**Pull request by sjavdani**
Added a new objective function, to be used for ComputeJointVelocityFromTwist, which adds a cost and gradient to avoid joint limits. Also modified ComputeJointVelocityFromTwist in order to make it have a hard constraint near a limit (was previously hard constrained at 0, now constrained at whatever joint_limit_tolerance is)

### CODE REVIEW

File: src/prpy/util.py [hunk 1]
```diff
@@ -428,11 +428,47 @@ def get_duration(self):
...
+    @param args[4] upper joint limit
+    '''
+    
+    objective, gradient = quadraticObjective(dq, *args)
+    q = args[2]
+    q_min = args[3]
+    q_max = args[4]
+
+    #when joint is less than this distance away from limit, start penalizing
+    delta_j = 3e-1
```
Comment (siddhss5) [inline src/prpy/util.py]: Why is this hardcoded?

File: src/prpy/util.py [hunk 2]
```diff
@@ -428,11 +428,47 @@ def get_duration(self):
...
+    qdiff_lower = delta_j - (q-q_min)
+    qdiff_upper = delta_j - (q_max-q)
+
+
+    dq_target = [diff_lower if diff_lower > 0. else
+                 -diff_upper if diff_upper > 0. else
+                 0. for diff_lower, diff_upper in zip(qdiff_lower, qdiff_upper)]
+
+    #constant for weighing cost of quadratic vs. being diff from bounds
+    lambda_dqdist = 0.01
```
Comment (siddhss5) [inline src/prpy/util.py]: Why is this hardcoded?

File: src/prpy/util.py [hunk 3]
```diff
@@ -446,7 +482,8 @@ def quadraticObjective(dq, *args):
 
 
 def ComputeJointVelocityFromTwist(robot, twist,
-                                  objective=quadraticObjective,
+                                  objective=quadraticPlusJointLimitObjective,
```
Comment (siddhss5) [inline src/prpy/util.py]: I think `quadraticObjective` should be default in `prpy.util`. There are way too many parameters in `quadraticPlusJointLimitObjective` for this to be easily generalizable to all robots. If you feel like it, you should use the other objective with specially tuned parameters within a specific `robot.py`.

File: src/prpy/util.py [hunk 4]
```diff
@@ -478,18 +515,20 @@ def ComputeJointVelocityFromTwist(robot, twist,
...
     if dq_init is None:
         dq_init = robot.GetActiveDOFVelocities()
 
     opt = scipy.optimize.fmin_l_bfgs_b(objective, dq_init, fprime=None,
-                                       args=(jacobian_active, twist_active),
+                                       args=(jacobian_active, twist_active, q_curr, q_min, q_max),
                                        bounds=dq_bounds, approx_grad=False)
 
     dq_opt = opt[0]
+#    if opt[1] > 0:
```
Comment (siddhss5) [inline src/prpy/util.py]: This has been removed in a downstream PR.

**Comment by siddhss5** [comment]
I'm concerned about this change. We use `VectorField` extensively and it's fast because `ComputeJointVelocityFromTwist` is fast. I'd be wary about adding more optimization logic. I understand stepwise performance can be better but I'm not sure it has been tested enough to warrant a wholesale switch.

**Comment by mkoval** [comment]
I do agree that we should leave `quadraticObjective` as the default and move this to `prpy.util`. It does take a lot of parameters that may not have sane default values.

**Comment by sjavdani** [comment]
Ok, changed the default to be quadratic objective. In order to make it so you can change the constants, I also changed it so some of the arguments had to be specified for both quadratic and quadratic+joint objectives.

Note that, compared to before, it has a different tolerance at which joint velocities are forced to be zero. I found this limit to work much better in practice - Sidd, if this messed up other stuff, feel free to make the constant (joint_limit_tolerance) smaller.

**Comment by mkoval** [comment]
Those changes look good to me. :+1: 

@siddhss5 Any other comments? Or can I merge?

**Pull request merged**